### PR TITLE
[CoqIDE] Cache the byte-unicode conversion inside the sentence type.

### PR DESCRIPTION
**Pull request by ppedrot**
Fixes the quadratic behaviour for good.

Alternative to #15798.

cc @SkySkimmer for 8.15.1 release.

### CODE REVIEW

**Comment by jfehrle** [comment]
What is the performance difference on Pff.v with this?

**Review by jfehrle — COMMENTED** [review]
minor comments

File: ide/coqide/coqOps.ml [hunk 1]
```diff
@@ -646,7 +695,11 @@ object(self)
     self#discard_command_queue queue;
     pop_info ();
     if Stateid.equal id tip || Stateid.equal id Stateid.dummy then begin
-      self#position_tag_at_iter ?loc start stop Tags.Script.error phrase;
+      let pos = match loc with
+      | None -> None
+      | Some loc -> Some (self#positions_of_loc sentence loc)
+      in
+      self#position_tag_at_iter ?pos start stop Tags.Script.error phrase;
```
[the remark this answers is not in the document]
Comment (ppedrot) [reply ide/coqide/coqOps.ml:702]: I could have used the map combinator but I don't like point-free style in OCaml. I don't think it hurts here to have twice a pattern matching. When a third one is introduced I'll reconsider.

**Comment by ppedrot** [comment]
> What is the performance difference on Pff.v with this?

Not sure how to reliably time this. I evaluated it several times, it didn't freeze and the hotspots I was seeing before with `perf` disappeared.

> Also, what's the performance if the user inserts a character at the beginning of the file just after CTRL-End?

It doesn't matter because the real problem is the quadratic behaviour. Introducing anything before the the current insertion point is O(n) because we recompute the list of sentences, so any additional O(n) is going to get drowned into the noise.

**Comment by jfehrle** [comment]
> Not sure how to reliably time this

I used my phone.

> > Also, what's the performance if the user inserts a character at the beginning of the file just after CTRL-End?

> It doesn't matter because the real problem is the quadratic behaviour. Introducing anything before the the current insertion point is O(n) because we recompute the list of sentences, so any additional O(n) is going to get drowned into the noise.

Nope.  I just tried to measure this case.  I gave up after 8 minutes.  The script was processed only up to line 10709 out of 27910.  With this PR and this test case, I think the conversion becomes inefficient both for unicode -> byte and vice-versa.

**Comment by ppedrot** [comment]
Wait a sec, now error highlighting is not performed correctly.

**Comment by ppedrot** [comment]
I really don't want to lose more time with CoqIDE so I am closing this for now. I think 8.15.1 can withstand a release in the current state, it's much better than 8.15.0 but there is something deeply broken in the location business that will need a complete rehauling.

**Comment by jfehrle** [comment]
The clean solution for 8.16 would be to change Loc.t to use unicode offsets rather than byte offsets.  The main change would be in the lexer.  This changes the semantics but not the syntax of some XML protocol messages.  It would require removing existing conversion code in CoqIDE and the other IDEs on the XML protocol.  We'd have to consider whether this would impact any plugins.

**Comment by ppedrot** [comment]
I think that it's an error to have absolute positions in locations, regardless of whether they're unicode or byte-based. This does not make sense in a non-linear document which is the kind of interaction we're currently promoting in Coq. We should instead have locations relative to the current sentence, as we were doing before, and store the sentence id in the location if we need to associate it with some physical mapping as text.

**Pull request closed without merging**